synapse/storage/_base.py

```diff
@@ -1216,3 +1217,46 @@ class _RollbackButIsFineException(Exception):
     something went wrong.
     """
     pass
+
+
+def db_to_json(db_content):
+    """
+    Take some data from a database row and return a JSON-decoded object.
+
+    Args:
+        db_content (memoryview|buffer|bytes|bytearray|unicode)
+    """
+    # psycopg2 on Python 3 returns memoryview objects, which we need to
+    # cast to bytes to decode
+    if isinstance(db_content, memoryview):
+        db_content = db_content.tobytes()
+
+    # psycopg2 on Python 2 returns buffer objects, which we need to cast to
+    # bytes to decode
+    if PY2 and isinstance(db_content, buffer):
+        db_content = bytes(db_content)
+
+    # Decode it to a Unicode string before feeding it to json.loads, so we
+    # consistenty get a Unicode-containing object out.
+    if isinstance(db_content, (bytes, bytearray)):
+        db_content = db_content.decode('utf8')
+
+    try:
+        return json.loads(db_content)
+    except Exception:
+        # I don't know how, but somehow some binary objects get mangled upon
```

Comment:
I'm assuming this is happening in the weird tables where we inexplicably store JSON as binary, like `users_filter`.  Looking at this with psql, the rows look like:

```
 matthew |       130 | \x7b22726f6f6d223a7b2274696d656c696e65223a7b226c696d6974223a32307d7d7d
```

which sure enough looks like \x7b ({) followed by a bunch of ASCII hex digits.

So, either it's being inserted in the first place in the mangled format, or the postgres client library (or server?) is incorrectly displaying it with only a single preceding \x rather than correctly putting a \x before each pair of hex digits.

The type of this column is:

`filter_json | bytea`

Looks like postgres is behaving as intended when displaying a bytea as `\x.............`; this is what you get if the bytea_output global is set to `hex`.  If you set it to `escape` (the other valid value), you get this instead:

```
matrix=# set bytea_output='escape';
SET
matrix=# SHOW bytea_output;
 bytea_output 
--------------
 escape
(1 row)

matrix=# select * from matrix.user_filters where user_id='matthew' limit 1;
 user_id | filter_id |             filter_json              
---------+-----------+--------------------------------------
 matthew |         0 | {"room": {"timeline": {"limit": 8}}}
(1 row)
```

So it looks like the data is okay in the DB, it's just that the python postgres library is interpreting `\x1234` as a python-style hex escape sequence rather than this horrible mutant postgres hex sequence.

So, looking at psycopg's FAQ (http://initd.org/psycopg/docs/faq.html#faq-bytea-9-0), it turns out we're not the only people with these woes, and the solution is simply to do the same trick as above and explicitly `set bytea_output='escape';` and then everything will be green & submarine.

tl;dr: let's kill this with fire and `set bytea_output='escape';` instead :)

Reply:
hm, actually, this might be insufficient - looking at https://github.com/cockroachdb/cockroach/pull/17693/files where they had a similar problem, they ended up manually unescaping it much as you have in the app :/

Reply:
actually, reading the FAQ some more, it sounds like this shouldn't even be a bug if we are running psycopg 2.4.1 or later with libpq from 9.0 or later.  On matrix.org right now we look to be running psycopg 2.7.5 with libpq 5.10 (i.e. for postgres 10) - so i think this should be absolutely fine.  Was this just that you were testing on a box using an ancient python postgres client?

I guess we should probably `set bytea_output='escape'` to be safe.

Reply:
(even nicer would be to kill these bytea tables with fire - i have **no idea** why they are byteas; it looks like they were originally created as LONGBLOBs in a fit of overenthusiasm/confusion, despite them just being utf8 JSON)

Reply:
@ara4n I noticed this coming out of *your* homeserver :)

Reply:
We should probably pin psycopg2 (and libpq, whose version psycopg2 helpfully exports). Debian jessie have suitably up to date packages, so it should be safe.